reindexer/src/main/scala/uk/ac/wellcome/platform/reindexer/services/ReindexTargetService.scala

```diff
@@ -0,0 +1,56 @@
+package uk.ac.wellcome.platform.reindexer.services
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
+import com.gu.scanamo.{Scanamo, Table}
+import com.gu.scanamo.error.DynamoReadError
+import com.gu.scanamo.query._
+import com.gu.scanamo.syntax._
+import uk.ac.wellcome.models.{Reindex, ReindexItem, Reindexable, Transformable}
+import uk.ac.wellcome.platform.reindexer.models.ReindexAttempt
+
+import scala.concurrent.Future
+import uk.ac.wellcome.utils.GlobalExecutionContext.context
+
+abstract class ReindexTargetService[T <: Transformable with Reindexable[
+  String]](dynamoDBClient: AmazonDynamoDB) {
+  import uk.ac.wellcome.utils.ScanamoUtils._
+
+  type ScanamoQuery =
+    (String, String) => (Query[_]) => List[Either[DynamoReadError, T]]
+
+  val transformableTable: Table[T]
+  val scanamoQuery: ScanamoQuery
+
+  private val gsiName = "ReindexTracker"
+
+  def runReindex(reindexAttempt: ReindexAttempt): Future[ReindexAttempt] =
+    for {
+      rows <- getRowsWithOldReindexVersion(reindexAttempt.reindex)
+      filteredRows = logAndFilterLeft(rows)
+      updateOps <- updateRows(reindexAttempt.reindex,
+                              filteredRows.map(_.getReindexItem))
+      updatedRows = logAndFilterLeft(updateOps)
+    } yield
+      reindexAttempt.copy(successful = updatedRows,
+                          attempt = reindexAttempt.attempt + 1)
+
+  def updateRows(reindex: Reindex, rows: List[ReindexItem[String]]) = {
```

Comment:
this should be private. If it's public for the sake of tests, then it should be split in its own class